### ISSUE TITLE
Unable to locate Ghidra's BuildStringModels class

### ISSUE THREAD

**Issue by pabx06** [NationalSecurityAgency/ghidra#2106]
hello. 
i would like to play around with the string Models of the StringsAnalyzer
I looked at this comment
https://github.com/NationalSecurityAgency/ghidra/blob/fe8d863c47f79d904c10c2c49d16ea4c1b674020/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/string/StringsAnalyzer.java#L43

But was unable to locate BuildStringModels class.
Any idea thanks.

**Comment by astrelsky** [comment]
I think this is now the class used to build the models.

https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/string/StringModel.java

**Comment by astrelsky** [comment]
> I think this is now the class used to build the models.
> 
> https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/string/StringModel.java

Also, this is great information to know because now I'm probably going to spend the day building some models.

**Comment by dragonmacher** [comment]
@astrelsky is correct.  The documentation is out-of-date.

**Comment by dragonmacher** [comment]
So, this is currently an immature feature.    It appears at first blush to be extensible, but we never provided the tools to allow users to change the functionality.   While it appears possible to change the model by using a different `.sng` file, we did not make it easy to do so.

We have not had many requests from users to update the string search model.   However,  being able to change the models would allow for improved string finding, especially for non-English languages.

To enable users to update the model we should at least provide scripts to build new models, along with help/documentation for using those scripts.

*Community feedback is needed for us assess the priority of improving the extensibility of this feature.*  **How important is this feature to you?**

**Comment by hedgeberg** [comment]
+1 to desiring to see the ability to add new String Models. Currently working on a binary which mixes ASCII, UTF-16, and EBCDIC character encodings, and being able to generate a string model for this environment is a very appealing idea. 

I think the biggest question for community utility is what would be needed to build these string models, and if changing the stock string model would change behavior in any unexpected ways.

**Comment by wyattearp** [comment]
I know this is an older thread, but I'd like to bump this as well. It looks like there's a set of tools missing based on the initial format of the file. Given the larger corpuses of data available now in comparision to only 4 years ago when the file was first release, being able generate a new string file specifically targeted towards contextual training data would great for UTF-16/32 software, specifically given that the Unicode BOM is optional and sadly not present most times.

**Comment by sad-dev** [comment]
Could I clarify how the commit introduces new string models? If I want to make a ``.sng`` model for say, Big5, how would I go about it?

**Comment by dev747368** [comment]
Re: trigrams, 3 "code points".  You are correct, the format of the .sng file is 3 characters(unicode code points) followed by the count for that trigram.  However, I suspect that a character set like big5 would produce more than 256 symbols and the symbol size value would need to be much larger.

Determining the symbol size value is not always obvious.  The value in the current sng file (128) isn't 100% accurate, but was chosen to match the pre-existing behavior.  The more correct value would be 128-26, because the string case is normalized ("Model Type: lowercase"), discarding 26 symbols ('A'-'Z') from the possible set of symbols that a trigram would be made of.

The threshold values will probably be hand-tuned by whoever is creating the model, feeding the training set or a testing set of values back through the model to see how many false positives / false negatives are produced.

On a related note, I have no idea if trigram probability scoring with a non-alphabetic language is as effective as what we get with our currently implemented ascii/latin/english model.  Would there need to be tweaks to the logic to support cleaning up values (like the current lower-casing option)?
